tools/resource-generator/resource-generator.go

```diff
@@ -79,6 +97,7 @@ func main() {
 	namespace := flag.String("namespace", "kube-system", "Namespace to use.")
 	pullPolicy := flag.String("pullPolicy", "IfNotPresent", "ImagePullPolicy to use.")
 	featureGates := flag.String("featureGates", "", "Feature gates to enable.")
+	infraReplicas := flag.String("infraReplicas", "2", "Number of replicas for virt-controller and virt-api")
```

Comment:
I'm wondering about the use case of the "infraReplicas" paramenter.

The number of virt-controller replicas only affects availability, but virt-api load balances work across replicas. So only the number of virt-api directly impacts performance

I understand that increasing the number of virt-api has its use case, but I don't see a reason why someone would want to increase both virt-controller and virt-api with the same number of replicas?

Reply:
The use-case I know of for this is actually single-node clusters that want only 1 replica of each.
However you are making a really good point that for other use-cases, tying virt-api and virt-controller together doesn't make that much sense...
Unfortunately, the main logic for this is already done (https://github.com/kubevirt/kubevirt/pull/6663)
This PR just surfaces an environment variable to make it easier to configure.
In the future, if we identify use-cases for 3+ replicas of either virt-api or virt-controller, we should probably add `virtApiReplicas` and `virtControllerReplicas` to the `infra` API (each would override `replicas` for their pods).

Reply:
Ok, It might be fine to change it later!

Reply:
I completely agree with @marceloamaral.

It may be too late, but TBH I think introducing such API option is problematic for the reasons specified above. There's an [ongoing effort](https://github.com/kubevirt/kubevirt/issues/7101) to investigate ways to auto-scale our infra components, and this option might interfere with that.

In general, my opinion always was that such decisions (at least at the vast majority of the cases) should not be controlled by the user. These kinds of decisions (such as determining infra components replicas) are hard to make, require deep understanding of the implementation and can change over time (for example let's say that we optimize `virt-api` so we would need only half of the replicas for large clusters - how would the user be aware of that?). In the extreme case of an advanced user that would like to change that we can always use `jsonpatch` in Kubevirt CR.

If it isn't too late I would remove this API from Kubevirt CR. At the very least I'd introduce documentation that says that this field should only be used by a very advanced user that knows what he's doing + that using this option prevents auto-scaling.

Sorry for not seeing your original PR earlier.

Reply:
> If it isn't too late I would remove this API from Kubevirt CR. At the very least I'd introduce documentation that says that this field should only be used by a very advanced user that knows what he's doing + that using this option prevents auto-scaling.

Unfortunately the API is already in 4 KubeVirt releases and can't just be removed.
However, I totally agree with a big warning about the option being for very advanced users only, I will add that to this PR.

We can also start thinking about deprecating the API to eventually remove it, but first we need a way to address single-replica use-cases.
The heuristic introduced in https://github.com/kubevirt/kubevirt/pull/7401 doesn't (yet) have a way to get only 1 replica.
This is where we need input from the user, because single-node users don't necessarily want single-replica configurations.
(And indeed, right now the default in kubevirtci is 1 node 2 replicas)
Maybe https://github.com/kubevirt/kubevirt/pull/7401 could introduce a `minReplicas` option that defaults to 2 but can be set to 1, and maybe even already mark `replicas` as deprecated?

Reply:
> I totally agree with a big warning about the option being for very advanced users only, I will add that to this PR.

Awesome! Thanks

> The heuristic introduced in https://github.com/kubevirt/kubevirt/pull/7401 doesn't (yet) have a way to get only 1 replica.

Come to think about it, a small improvement to the heuristic might be good enough - simply scale down to 1 replica if it's a single-node cluster. WDYT?

> We can also start thinking about deprecating the API to eventually remove it

I think the biggest problem is that we don't have a deprecation policy. But sure, I'm all for it if it's possible!

Reply:
> Come to think about it, a small improvement to the heuristic might be good enough - simply scale down to 1 replica if it's a single-node cluster. WDYT?

I think that would be great, but keep in mind it would be a radical change, as it would affect the default behavior of kubevirtci deployments.

Reply:
> I think that would be great, but keep in mind it would be a radical change, as it would affect the default behavior of kubevirtci deployments.

Why is that? AFAIK, kubevirtCI clusters are deployed with 3 nodes. Am I missing something?

> Do you think it would make sense to add the warning as part of https://github.com/kubevirt/kubevirt/pull/7401 instead?

Yes, makes a lot of sense.
BTW, in which form did you think on providing this warning? By logging (`log.Log.Warning(...)`)?

Reply:
> > I think that would be great, but keep in mind it would be a radical change, as it would affect the default behavior of kubevirtci deployments.
> 
> Why is that? AFAIK, kubevirtCI clusters are deployed with 3 nodes. Am I missing something?
> 

The default number of nodes is 1. Most, if not all test lanes override it to 3, but dev environments usually don't bother.

Reply:
Oh, right. I do overwrite it by default, so forgot about that fact. In any case I think this shouldn't be a problem.. right? :sweat_smile:

Reply:
🤞

Reply:
> > Do you think it would make sense to add the warning as part of #7401 instead?
> 
> Yes, makes a lot of sense. BTW, in which form did you think on providing this warning? By logging (`log.Log.Warning(...)`)?

I would add it first to the API description, so users are warned ahead of time. But yeah, adding a warning can't hurt!

Reply:
I think having an option to say: Never scale above this limit for our components is fine (maybe individual limits may be preferable).

In general I agree with all here that we are not in an optimal spot right now. We definitely want  support for

 1. autoscaling virt-api and
 1. scaling with the k8s-apiserver (basically with control-plane nodes) (wanted to implement this via a daemonset in the past)

Especially scaling with `control-plane` nodes is in my opinion the solution which gives almost always the right amount of  instances. It considers availability (for both virt-api and virt-controller) as well as load-scaling (we scale together with the apiserver. The apiserver proxies all our long-living connections as well, so has always more load than we have. So just following the apiserver sounds like a good default heuristic).

For free auto-scaling I think we should definitely have a way to specify something like a max-amount of replicas instead of a fixed amount.